### DISCUSSION TITLE
Enhancement: View Actions

### DISCUSSION THREAD

**Discussion by hasansezertasan** [jowilf/starlette-admin#412] [Ideas]
**Is your feature request related to a problem? Please describe.**
Having batch actions and row actions is awesome for the record based actions but what if I want to add custom create function with less required fields.

Other scenario could be a batch action for the "entire table", which is lacking us because of the pagination. I may want to apply a batch action to the entire table, what should I do? Paginate the entire table and repeat the same action for all?

Another scenario might only be a functionality that is not related to any record. Starlette Admin has a lot of potential to be it's own web framework and offering features like this could attract the developers (like me)...

The buttons could be above or below the table or maybe in the menu?

**Describe alternatives you've considered**
FastAPI Amis Admin is quite good for configurations like this but it doesn't use Jinja2 and it lacks English documentation 

**Additional context**
I'm planning to do some work on this, wdyt about this feature @jowilf?

**Comment by mrharpo** [comment]
+1 for this feature

We needed this recently for a bulk csv uploader, and I was able to make it pretty simply by overwriting the list view template and manually adding the button I need. [Here's the full implementation](https://github.com/WGBH-MLA/organ/pull/14)

But in short:

## View
```py
class OpenVaultCatalogView(ModelView):
    label = 'Open Vault Catalog'
    list_template = 'custom-list-view.html'

    # This is to prevent the upload_csv action from being shown in the actions dropdown when items are selected
    actions = ['delete']

    # Create a regular action
    @action(
        name='upload_csv',
        text='Upload CSV',
        custom_response=True,
    )
    # pks required for the function signature, but aren't used for this logic
    async def upload_csv(self, request, pks):
        form = await request.form()
        # business logic
        ...
        # If something goes wrong, let the user know
        raise ActionFailed(f'error ingesting: {errors}')
        # Otherwise, show a success message
        return f'Successfully ingested {n} records'
```

## Template
```jinja
<!-- custom-list-view.html -->
                        ...
                        <div class="card-actions">
                            <div class="btn-list">
                                {% include "actions.html" %}

                                {% if model.can_create(request) %}

                                   <!-- Add the custom action button -->
                                    <a title="Upload CSV"  class="btn btn-secondary btn-block ms-2" href="#" data-bs-toggle="modal" data-bs-target="#modal-action" data-name="upload_csv" data-confirmation="Please make sure the CSV file is formatted correctly with no headers: OVID,GUID" data-form="<form id=&quot;modal-form&quot; name=&quot;upload_csv&quot; method=&quot;POST&quot; enctype=&quot;multipart/form-data&quot;>
                                        <input type=&quot;file&quot; name=&quot;csv_file&quot; accept=&quot;.csv&quot; required>
                                        </form>" data-submit-btn-class="btn-primary">
                                        <i class="fa-solid fa-upload me-2"></i>
                                        Upload CSV
                                    </a>
                                    ...
<!-- rest of the template -->
```


## Results
![image](https://github.com/user-attachments/assets/b469b555-3425-4110-ac95-477e2ce6f35e)

![image](https://github.com/user-attachments/assets/7c566187-737f-4678-b423-e6b870098c77)

![image](https://github.com/user-attachments/assets/3f99dc64-c1a5-45c7-b3e6-5b9a4fbf0030)

## Feature request
It would be really nice to have this as a decorator like `@view_action`, since all the necessary form values can already be set inside the existing `@action` and `@row_action` signatures, and the custom template wouldn't be needed.
